File: client/src/hexmath.ts
import { HEX_SIZE } from './constants';

export interface Hex {
  q: number;
  r: number;
}

const DIRECTIONS: Hex[] = [
  { q: 1, r: 0 }, { q: 1, r: -1 }, { q: 0, r: -1 },
  { q: -1, r: 0 }, { q: -1, r: 1 }, { q: 0, r: 1 },
];

export function neighbors(h: Hex): Hex[] {
  return DIRECTIONS.map(d => ({ q: h.q + d.q, r: h.r + d.r }));
}

export function hexToPixel(h: Hex): { x: number; y: number } {
  const x = HEX_SIZE * (Math.sqrt(3) * h.q + (Math.sqrt(3) / 2) * h.r);
  const y = HEX_SIZE * (1.5 * h.r);
  return { x, y };
}

export function pixelToHex(px: number, py: number): Hex {
  const q = (Math.sqrt(3) / 3 * px - 1 / 3 * py) / HEX_SIZE;
  const r = (2 / 3 * py) / HEX_SIZE;
  return hexRound(q, r);
}

function hexRound(q: number, r: number): Hex {
  const s = -q - r;
  let rq = Math.round(q);
  let rr = Math.round(r);
  const rs = Math.round(s);

  const qDiff = Math.abs(rq - q);
  const rDiff = Math.abs(rr - r);
  const sDiff = Math.abs(rs - s);

  if (qDiff > rDiff && qDiff > sDiff) {
    rq = -rr - rs;
  } else if (rDiff > sDiff) {
    rr = -rq - rs;
  }

  return { q: rq, r: rr };
}

export function hexKey(h: Hex): string {
  return `${h.q},${h.r}`;
}

File: internal/game/hexmath.go
package game

import "math"

type Hex struct {
	Q int `json:"q"`
	R int `json:"r"`
}

var axialDirections = [6]Hex{
	{1, 0}, {1, -1}, {0, -1},
	{-1, 0}, {-1, 1}, {0, 1},
}

func (h Hex) Neighbors() [6]Hex {
	var out [6]Hex
	for i, d := range axialDirections {
		out[i] = Hex{h.Q + d.Q, h.R + d.R}
	}
	return out
}

func (h Hex) IsNeighbor(other Hex) bool {
	dq := h.Q - other.Q
	dr := h.R - other.R
	ds := (-h.Q - h.R) - (-other.Q - other.R)
	return (abs(dq)+abs(dr)+abs(ds))/2 == 1
}

func (h Hex) Distance(other Hex) int {
	dq := h.Q - other.Q
	dr := h.R - other.R
	ds := (-h.Q - h.R) - (-other.Q - other.R)
	return (abs(dq) + abs(dr) + abs(ds)) / 2
}

const hexSize = 30.0

func HexToPixel(h Hex) (float64, float64) {
	x := hexSize * (math.Sqrt(3)*float64(h.Q) + math.Sqrt(3)/2*float64(h.R))
	y := hexSize * (3.0 / 2 * float64(h.R))
	return x, y
}

func PixelToHex(x, y float64) Hex {
	q := (math.Sqrt(3)/3*x - 1.0/3*y) / hexSize
	r := (2.0 / 3 * y) / hexSize
	return hexRound(q, r)
}

func hexRound(q, r float64) Hex {
	s := -q - r
	rq := math.Round(q)
	rr := math.Round(r)
	rs := math.Round(s)

	qDiff := math.Abs(rq - q)
	rDiff := math.Abs(rr - r)
	sDiff := math.Abs(rs - s)

	if qDiff > rDiff && qDiff > sDiff {
		rq = -rr - rs
	} else if rDiff > sDiff {
		rr = -rq - rs
	}

	return Hex{int(rq), int(rr)}
}

func abs(x int) int {
	if x < 0 {
		return -x
	}
	return x
}
